src/Diagnostics/DiagnosticsMachine/DiagnosticsMachine.jl

```diff
@@ -0,0 +1,112 @@
+"""
+    DiagnosticsMachine
+
+This module provides the infrastructure to extract diagnostics from a
+ClimateMachine simulation. Two key abstractions are defined: diagnostic
+variables and diagnostic groups. The `StdDiagnostics` module makes use of
+these to define many standard variables and groups which may be used
+directly by experiments. `DiagnosticsMachine` may be used by experiments
+to define specialized variables and groups.
+"""
+module DiagnosticsMachine
+
+export DiagnosticVar,
+    PointwiseDiagnostic,
+    @pointwise_diagnostic,
+    dv_PointwiseDiagnostic,
+    HorizontalAverage,
+    @horizontal_average,
+    dv_HorizontalAverage,
+    ScalarDiagnostic,
+    @scalar_diagnostic,
+    dv_ScalarDiagnostic,
+    States,
+    #DiagnosticsGroup,
+    @diagnostics_group
+#DiagnosticsGroupParams
+
+using CUDA
+using Dates
+using InteractiveUtils
+using KernelAbstractions
+using MacroTools
+using MacroTools: prewalk
+using MPI
+using OrderedCollections
+using Printf
+using StaticArrays
+
+using ..Diagnostics # temporarily, for DiagnosticsGroup
+using ..Atmos
+using ..BalanceLaws
+using ..ConfigTypes
+using ..DGMethods
+using ..GenericCallbacks
+using ..Mesh.Grids
+using ..Mesh.Interpolation
+using ..Mesh.Topologies
+using ..MPIStateArrays
+using ..Spectra
+using ..TicToc
+using ..VariableTemplates
+using ..Writers
+
+using CLIMAParameters
+using CLIMAParameters.Planet: planet_radius
+
+# Container to store simulation information necessary for all
+# diagnostics groups.
+Base.@kwdef mutable struct Diagnostic_Settings
+    mpicomm::MPI.Comm = MPI.COMM_WORLD
+    param_set::Union{Nothing, AbstractParameterSet} = nothing
+    dg::Union{Nothing, SpaceDiscretization} = nothing
+    Q::Union{Nothing, MPIStateArray} = nothing
+    starttime::Union{Nothing, String} = nothing
+    output_dir::Union{Nothing, String} = nothing
+    no_overwrite::Bool = false
+end
+const Settings = Diagnostic_Settings()
+
+include("onetime.jl")
+include("variables.jl")
+include("groups.jl")
+
+const AllDiagnosticVars = OrderedDict{
+    Type{<:ClimateMachineConfigType},
+    OrderedDict{String, DiagnosticVar},
+}()
```

Comment:
I agree in principle, but don't see a way to get rid of this -- it is used when creating a diagnostics group, to look up the diagnostic variables declared to be in the group.

Reply:
Trying to remove this, I find myself in scope resolution hell. We have diagnostic variables defined in the `StdDiagnostics` module, but we can also have them in user code. This table stores instances of all diagnostic variables, irrespective of where they're defined, and without it it is difficult to generate an instance of a diagnostic variable given just the name since we don't know in which module it was defined.